app/app/dialog/[id]/page.tsx

```diff
@@ -1,16 +1,241 @@
+              <div className="flex flex-col max-w-[75%]">
+                {showSenderHeader && (
+                  <div className="flex items-center gap-2 mb-2">
+                    <Avatar>
+                      <AvatarImage src={thumbSrc} /> 
+                      <AvatarFallback>{getInitials(sender)}</AvatarFallback>
+                    </Avatar>
+                    <p className="text-xs text-muted-foreground font-medium">
+                      {sender}
+                    </p>
+                  </div>
+                )}
+                <div
+                  className={`px-4 py-2 text-sm rounded-xl whitespace-pre-line shadow-sm break-words ${
+                    isOutgoing
+                      ? 'bg-blue-500 text-white rounded-br-none'
+                      : 'bg-gray-100 text-foreground rounded-bl-none'
+                  }`}
+                >
+                  {msg.message ? 
+                    (<p>{msg.message}</p>) :
+                    (<Media  />)
+                  }
+                  <div className={`text-xs text-right mt-1 ${
+                    isOutgoing
+                      ? 'text-gray-300'
+                      : 'text-muted-foreground'
+                    }`}>
+                    {formatDate(msg.date)}
+                  </div>
+                </div>
+              </div>
+            </div>
+          )
+        })}
+      </div>
+    </div>
+  )
+}
+
+const getFromStoracha = async (cid: string) => {
+  const url = new URL(`/ipfs/${cid}`, process.env.NEXT_PUBLIC_STORACHA_GATEWAY_URL)
+  const response = await fetch(url)
+  if (!response.ok) {
+    throw new Error(`Failed to fetch: ${response.status} ${response.statusText} ${url}`);
+  }
+  return response
+}
+
 export default function Page () {
+  const [{ client }] = useTelegram()
+  const [loading, setLoading] = useState(true)
+  const [dialogData, setDialogData] = useState<DialogData | null>(null)
+  const [messages, setMessages] = useState<MessageData[]>([])
+  const [userId, setUserId] = useState<string>()
+  const [participants, setParticipants] = useState<Record<string, EntityData>>({})
   const params = useParams<{ id: string }>()
-  console.log(params)
+  const searchParams = useSearchParams()
+  const backupCid = searchParams.get('backupData')
+
+  useEffect(() => {
+      const restoreBackup = async () => {
+        try{
+          if (!client.connected) await client.connect()
+
+          const userId = (await client.getMe()).id
+          setUserId(userId.toString())
+
+          if(!cloudStorage.getKeys.isAvailable()){
+            throw new Error('Error trying to access cloud storage')
+          }
+          
+          const encryptionPassword = await cloudStorage.getItem('encryption-password')
+          if (!encryptionPassword) {
+            throw new Error('Encryption password not found in cloud storage')
+          }
+          
+          const response = await getFromStoracha(`${backupCid}?format=car`)
```

Comment:
I believe this will actually retrieve a CAR with _all_ the blocks in the backup since it traverses the DAG and you're not specifying a "dag-scope".

You're then proceeding to make separate requests for the other data you need (but actually already have)...

You should use `format=raw` to get the root block - then you don't need to use the CAR decoder, you can just `dagCBOR.decode(new Uint8Array(await response.arrayBuffer()))`.